run/run.go

```diff
@@ -1,47 +1,154 @@
+	}
+	err = json.Unmarshal(docB, &doc)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshall vulnerabilities document: %v", err)
 	}
 
-	if len(image.FsLayers) == 0 {
-		return nil, nil, fmt.Errorf("failed to pull pull fsLayers")
+	commands, err := GetImageCommands(conf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get image commands : %v", err)
+	}
+
+	return &doc, commands, nil
+}
+
+func ExecuteLocalGrypeScan(imageName string, conf *config.Config) (*grype_models.Document, []*docker.FsLayerCommand, error) {
+	dbConfig := grype_db.Config{
+		DBRootDir:           "/tmp/",
+		ListingURL:          "https://toolbox-data.anchore.io/grype/databases/listing.json",
+		ValidateByHashOnGet: false,
+	}
+	provider, metadataProvider, dbStatus, err := grype.LoadVulnerabilityDB(dbConfig, true)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load vulnerability DB: %w", err)
+	}
+	if dbStatus == nil {
+		return nil, nil, fmt.Errorf("unable to determine DB status")
 	}
 
-	commands := image.GetFsCommands()
+	registryOptions := &anchore_image.RegistryOptions{
+		InsecureSkipTLSVerify: conf.DockerConfig.InsecureTLS,
+		InsecureUseHTTP:       conf.DockerConfig.InsecureRegistry,
+		Credentials: []anchore_image.RegistryCredentials{
+			{
+				Authority: "", // What is this?
+				Username:  conf.DockerConfig.User,
+				Password:  conf.DockerConfig.Password,
+				Token:     conf.DockerConfig.Token,
+			},
+		},
+	}
+	packages, context, err := grype_pkg.Provide(imageName, source.SquashedScope, registryOptions)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to analyze packages: %v", err)
+	}
 
-	log.Infof("Analysing %d layers", len(image.FsLayers))
+	allMatches := grype.FindVulnerabilitiesForPackage(provider, context.Distro, packages...)
 
-	var vulnerabilities []*clair.Vulnerability
+	doc, err := grype_models.NewDocument(packages, context, allMatches, nil, metadataProvider, nil, dbStatus)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create document: %v", err)
+	}
 
-	c := clair.NewClair(conf.ClairAddr, conf.ClairTimeout)
-	vulnerabilities, err = c.Analyse(image)
+	commands, err := GetImageCommands(conf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get image commands : %v", err)
+	}
+
+	return &doc, commands, nil
+}
+
+func createGrypeClient(serverAddress string) *grype_client.GrypeServer {
+	cfg := grype_client.DefaultTransportConfig()
+	cfg.WithHost(serverAddress)
+	transport := httptransport.New(cfg.Host, cfg.BasePath, cfg.Schemes)
+	return grype_client.New(transport, strfmt.Default)
+}
+
+func GetImageCommands(conf *config.Config) ([]*docker.FsLayerCommand, error) {
+	image, err := docker.NewImage(&conf.DockerConfig)
 	if err != nil {
-		log.Errorf("Failed to analyze using API: %s", err)
-	} else {
-		if !conf.JSONOutput {
-			log.Infof("Got results from Clair API")
-		}
+		return nil, fmt.Errorf("failed to parse name: %v", err)
+	}
+
+	err = image.Pull()
```

Comment:
so we basically "pull" the image twice, one in the scanning and one for the image commands?

Reply:
It was always like this. it's not really pull of the tar, it's just the manifest. Didn't want to touch the commands logic part of this task.